File: java/homework_oop_3/src/food/Spaghetti.java
package food;

public class Spaghetti extends Food {
	private double _pasta;
	private double _sauce;
	private double _meat;
	private double _cheese;
	private boolean _cached = false;
	private double _calories = Double.NaN;
	
	public Spaghetti(double pasta, double sauce, double meat, double cheese) {
		_pasta = pasta;
		_sauce = sauce;
		_meat = meat;
		_cheese = cheese;
	}
	
	private void updateCache() {
		if (!_cached) {
			_calories = _pasta * 4 + _sauce * 3 + _meat * 100 + _cheese * 70;
			_cached = true;
		}
		
	}
	
	public double getPasta() { return _pasta; }
	public void setPasta(double value) {
		if (value != _pasta) {
			_pasta = value;
			_cached = false;
		}
	}
	
	public double getSauce() { return _sauce; }
	public void setSauce(double value ) {
		if(value != _sauce) {
			_sauce = value;
			_cached = false;
		}
	}
	
	public double getMeat() { return _meat;}
	public void setMeat(double value) {
		if (value != _meat) {
			_meat = value;
			_cached = false;
		}
	}

	public double getCheese() { return _cheese; }
	public void setCheese(double value) {
		if (value != _cheese) {
			_cheese = value;
			_cached = false;
		}
	}
	
	public double getCalories() {
		if (!_cached) updateCache();
		return _calories;
	}
	
	public void eat() { System.out.println("Eatin spaghetti!"); }
}

File: cpp/homework_oop_4/src/Spaghetti.cpp
#include "Spaghetti.h"

namespace foods {
  double Spaghetti::getMeat() const {
    return _meat;
  }
  void Spaghetti::setMeat(const double & val) {
    if (_meat != val){
      _meat = val;
      _cached = false;
    }
  }

  double Spaghetti::getCheese() const {
    return _cheese;
  }
  void Spaghetti::setCheese(const double &val) {
    if (_cheese != val) {
      _cheese = val;
      _cached = false;
    }
  }

  double Spaghetti::getSauce() const {
    return _sauce;
  }
  void Spaghetti::setSauce(const double &val) {
    if (_sauce != val) {
      _sauce = val;
      _cached = false;
    }
  }

  void Spaghetti::updateCalories() {
    _calories = 23*_meat + 42*_cheese + 34*_sauce;
  }

  double Spaghetti::getCalories() {
    if (!_cached) updateCalories();
    return _calories;
  }
  Spaghetti::Spaghetti(const double &meat, const double &cheese, const double &sauce)
    : _meat(meat), _cheese(cheese), _sauce(sauce), _calories(23*meat + 42*cheese + 34*sauce), _cached(true)
  {
  }

}
